### DISCUSSION TITLE
Expose Kratos as a http.Handler

### DISCUSSION THREAD

**Discussion by krhubert** [ory/kratos#1746] [Ideas]
Hi,

It would be great to be able to mount Kratos `http.Handler` inside app without setting up a microservice.

**Comment by aeneasr** [comment]
There are unfortunately no plans to support that as there is really a lot of bootstrapping required. Also, we do not want to maintain API compatibility for internal APIs as it is a lot of unnecessary work and the gained benefit is small!

**Reply by krhubert** [reply]
Thanks for the quick response!

**Comment by zepatrik** [comment]
I like the idea of importing the cobra commands and running them directly within go, similar to https://github.com/ory/keto/blob/fe92134db8c22b6c7c2b11397d8c1dc0e36c12ae/internal/e2e/cli_client_test.go
What do you think about that @aeneasr ?
Some stuff could then be passed through the context, e.g. by adding a logger to the context. We would then have a public go API, without the need to expose internals. The commands are quite stable and if they change it is a breaking change anyway as you have to adjust e.g. scripts or docker files anyway.

This would look similar to
```go
func main() {
	l := myLogrusLogger()
	ctx := kratos.WithLogger(context.Background(), l)
	ctx = kratos.WithConfigValue(ctx, "public.serve.port", 0)

	k := kratos.RootCommand()
	k.SetArgs([]string{"serve", "-c", myConfigPath)

	k.ExecuteContext(ctx)
}
```

See https://github.com/ory/x/blob/0b4ae5055bc414c834728232b3cfbfb07b7c1d10/cmdx/helper.go#L129-L147 for more examples on how cobra commands can be set up and run programmatically.

**Comment by krhubert** [comment]
> Also, we do not want to maintain API compatibility for internal APIs

Also quick note. What about exposing API without a compatibility guarantee. From what I've checked the bootstrapping is not that hard as it is mainly create a driver in [serve/root.go](https://github.com/ory/kratos/blob/master/cmd/serve/root.go) and then a bit setup in [serve.go](https://github.com/ory/kratos/blob/master/cmd/daemon/serve.go).

The `ServePublic` cloud be exposed as a function and return httprouter Handler. 

I'm not saying about allowing some crazy integration like choosing a Postgres driver, logging lib etc. Just move HTTP handler creation to separate function and that's it (without guarantee, nothing). Use it at your own risk.

**Reply by zepatrik** [reply]
That would not be enough IMO, as you have to at least run SQL migrations additionally. Possibly some more things that the standalone Kratos can do (e.g. email courier). Would that be included in the handler? If so, it would have to be configurable probably, and many more to come...
